core/src/main/java/org/elasticsearch/action/ingest/IngestProxyActionFilter.java

```diff
@@ -0,0 +1,152 @@
+            IndexRequest indexRequest = (IndexRequest) request;
+            isIngestRequest = Strings.hasText(indexRequest.pipeline());
+        } else if (BulkAction.NAME.equals(action)) {
+            ingestAction = BulkAction.INSTANCE;
+            assert request instanceof BulkRequest;
+            BulkRequest bulkRequest = (BulkRequest) request;
+            for (ActionRequest actionRequest : bulkRequest.requests()) {
+                if (actionRequest instanceof IndexRequest) {
+                    IndexRequest indexRequest = (IndexRequest) actionRequest;
+                    if (Strings.hasText(indexRequest.pipeline())) {
+                        isIngestRequest = true;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (isIngestRequest) {
+            assert ingestAction != null;
+            forwardIngestRequest(ingestAction, request, listener);
+            return;
+        }
+        chain.proceed(task, action, request, listener);
+    }
+
+    private void forwardIngestRequest(Action action, ActionRequest request, ActionListener listener) {
+        transportService.sendRequest(randomIngestNode(), action.name(), request, new TransportResponseHandler<TransportResponse>() {
+            @Override
+            public TransportResponse newInstance() {
+                return action.newResponse();
+            }
+
+            @Override
+            @SuppressWarnings("unchecked")
+            public void handleResponse(TransportResponse response) {
+                listener.onResponse(response);
+            }
+
+            @Override
+            public void handleException(TransportException exp) {
+                listener.onFailure(exp);
+            }
+
+            @Override
+            public String executor() {
+                return ThreadPool.Names.SAME;
+            }
+        });
+    }
+
+    @Override
+    public void apply(String action, ActionResponse response, ActionListener listener, ActionFilterChain chain) {
+        chain.proceed(action, response, listener);
+    }
+
+    @Override
+    public int order() {
+        return Integer.MAX_VALUE;
+    }
+
+    private DiscoveryNode randomIngestNode() {
+        assert NodeModule.isNodeIngestEnabled(clusterService.localNode().attributes()) == false;
+        List<DiscoveryNode> ingestNodes = new ArrayList<>();
+        for (DiscoveryNode node : clusterService.state().nodes()) {
+            if (NodeModule.isNodeIngestEnabled(node.getAttributes())) {
+                ingestNodes.add(node);
+            }
+        }
+
+        if (ingestNodes.isEmpty()) {
+            throw new IllegalStateException("There are no ingest nodes in this cluster, unable to forward request to an ingest node.");
+        }
+
+        int index = getNodeNumber();
+        return ingestNodes.get((index) % ingestNodes.size());
+    }
+
+    private int getNodeNumber() {
+        int index = randomNodeGenerator.incrementAndGet();
```

Comment:
If `randomNodeGenerator` would go over `Integer.MAX_VALUE` then we will return the same node number until `randomNodeGenerator` is back to value `1`? So we're better off with the current code?